Fix CID #102150

JackDanger opened the pull request:
Don't attempt to write an error message to a channel that may have been
closed and freed.

[CID #102150](https://scan8.coverity.com/reports.htm#v22612/p10672/fileInstanceId=3625286&defectInstanceId=1525721&mergedDefectId=102150)

JackDanger committed:
Fix CID #102150

Don't attempt to write an error message to a channel that may have been
closed and freed.

[CID #102150](https://scan8.coverity.com/reports.htm#v22612/p10672/fileInstanceId=3625286&defectInstanceId=1525721&mergedDefectId=102150)

equalsraf commented:
LGTM

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -442,17 +442,22 @@ static void handle_request(Channel *channel, msgpack_object *request)
+                                                    &out_buffer));
+
+    // Checking the result of the above write because the channel may
+    // be closed and freed.
```
This is a somewhat redundant comment given that `if (success)` is pretty self-documenting. In most of the source, the pattern is usualy `if(channel_write(...))`, so the presence of `success` here is a luxury already. I think the comment can be removed, or at least trimmed to 1 line.

JackDanger replied:
I've swapped out the `success` boolean for the `if(channel_write(` approach you mentioned. I think it looks much clearer now.

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -619,8 +621,7 @@ static void unsubscribe(Channel *channel, char *event)
 
-/// Close the channel streams/job. The channel resources will be freed by
-/// free_channel later.
+/// Close the channel streams/job and free the channel memory.
```
Since you changed the comment: "resources" is more accurate than "memory" because it also closes the libuv handle.

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -442,17 +442,19 @@ static void handle_request(Channel *channel, msgpack_object *request)
+                                        &error,
+                                        NIL,
+                                        &out_buffer))) {
+      // And close the channel
```
IMO, the error message on line 454 makes this comment redundant.

JackDanger replied:
Ha! I totally didn't notice that - you're completely right. Fixing.

JackDanger commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -619,8 +620,7 @@ static void unsubscribe(Channel *channel, char *event)
 
-/// Close the channel streams/job. The channel resources will be freed by
-/// free_channel later.
+/// Close the channel streams/job and free the channel resources.
```
@justinmk You're totally right. Reverting to 'resources'

justinmk commented:
@JackDanger Sorry for the friction on this PR, there's just one little clint complaint left. Can you fix it and rebase?

JackDanger commented:
> Sorry for the friction on this PR, there's just one little clint complaint left. Can you fix it and rebase?

No apologies necessary! I feel like I'm learning a lot from all the help. Fixed and rebased.

justinmk commented:
Thanks again. For future coverity PRs, try to follow the `coverity/<id>` convention mentioned here: https://github.com/neovim/neovim/wiki/Contributing#coverity-conventions
